src/User32.Desktop/User32.cs

```diff
@@ -621,10 +674,478 @@ public static partial class User32
+        /// <param name="lParam">An application-defined value to be passed to the callback function.</param>
+        /// <returns>
+        /// If the function is unable to perform the enumeration, the return value is zero. Call GetLastError to get extended error information.
+        /// If the function succeeds, it returns the nonzero value returned by the callback function that was pointed to by <paramref name="lpEnumFunc"/>.
+        /// If the callback function fails, the return value is zero. The callback function can call SetLastError to set an error code for the caller to retrieve by calling GetLastError.
+        /// </returns>
+        /// <remarks>
+        /// The EnumWindowStations function enumerates only those window stations for which the calling process has the <see cref="WINDOW_STATION_ACCESS_MASK.SpecificRight.WINSTA_ENUMERATE"/> access right.
+        ///  EnumWindowStations repeatedly invokes the <paramref name="lpEnumFunc"/> callback function until the last window station is enumerated or the callback function returns FALSE.
+        /// </remarks>
+        [DllImport(nameof(User32), SetLastError = true)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        public static extern bool EnumWindowStations(EnumWindowStationDelegate lpEnumFunc, IntPtr lParam);
+
+        /// <summary>
+        /// Enumerates all window stations in the current session. The function passes the name of each window station, in turn, to an application-defined callback function.
+        /// </summary>
+        /// <param name="lpEnumFunc">An application-defined <see cref="EnumWindowStationDelegate"/> callback function.</param>
+        /// <param name="lParam">An application-defined value to be passed to the callback function.</param>
+        /// <returns>
+        /// If the function is unable to perform the enumeration, the return value is zero. Call GetLastError to get extended error information.
+        /// If the function succeeds, it returns the nonzero value returned by the callback function that was pointed to by <paramref name="lpEnumFunc"/>.
+        /// If the callback function fails, the return value is zero. The callback function can call SetLastError to set an error code for the caller to retrieve by calling GetLastError.
+        /// </returns>
+        /// <remarks>
+        /// The EnumWindowStations function enumerates only those window stations for which the calling process has the <see cref="WINDOW_STATION_ACCESS_MASK.SpecificRight.WINSTA_ENUMERATE"/> access right.
+        ///  EnumWindowStations repeatedly invokes the <paramref name="lpEnumFunc"/> callback function until the last window station is enumerated or the callback function returns FALSE.
+        /// </remarks>
+        [DllImport(nameof(User32), SetLastError = true)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        public static extern bool EnumWindowStations(IntPtr lpEnumFunc, IntPtr lParam);
```

Comment:
Yes, I am aware of that. I have been maintaining a WindowsHook port for managed code. By the way I want to switch to our PInvoke libraries instead of P/Invok'ing them by myself. But to do that we would need to change our SetWindowHook callback and I would like to extend it to use the proper callback types for each type of hook, as it is meant to be. The one we have implemented is a very generic implementation of the specialized callback functions. I guess that might be something to discuss on issues also.

I had to create a unmanaged library to be able to hook on all kinds of hooks because of that problem. Some hooks do work fine under managed code, but some special cases and all these out-of-process hooks don't. That always makes me wishing for reverse P/Invoke on .NET platform ;-)

Something to discuss with the CoreCLR guys probably. Specially now that .NET is getting more and more prevalent on other platforms. Needed this feature about three or four different times on different projects on the last year.